xla/service/gpu/stream_attribute_annotator.cc

```diff
@@ -130,26 +131,45 @@ absl::StatusOr<bool> StreamAttributeAnnotator::Run(
   XLA_VLOG_LINES(
       5, "StreamAttributeAnnotator::Run(), before:\n" + module->ToString());
   bool changed = false;
+  int64_t host_memory_color =
+      static_cast<int64_t>(stream_executor::MemoryType::kHost);
   for (const HloComputation* comp : module->computations(execution_threads)) {
     for (HloInstruction* instr : comp->MakeInstructionPostOrder()) {
       auto instr_gpu_config = instr->backend_config<GpuBackendConfig>();
       if (!instr_gpu_config.ok()) {
         continue;
       }
-      // For fusion instruction, only annotate
-      // when the root of fusion is a single instruction
-      // running on non-default stream.
-      if (instr->opcode() == HloOpcode::kFusion) {
-        TF_ASSIGN_OR_RETURN(bool comp_result,
-                            AnnotateStreamAttributesForInstruction(
-                                instr, instr_gpu_config.value()));
-        changed |= comp_result;
+      if (!copy_start_done_) {
+        // For fusion instruction, only annotate
+        // when the root of fusion is a single instruction
+        // running on non-default stream.
+        if (instr->opcode() == HloOpcode::kFusion) {
+          TF_ASSIGN_OR_RETURN(bool comp_result,
+                              AnnotateStreamAttributesForInstruction(
+                                  instr, instr_gpu_config.value()));
+          changed |= comp_result;
+        }
+
+        TF_ASSIGN_OR_RETURN(
+            bool user_result,
+            AnnotateStreamAttributesForUsers(instr, instr_gpu_config.value()));
+        changed |= user_result;
+      } else {
+        if (instr->opcode() == HloOpcode::kCopyStart) {
```

Comment:
i think if you just move this before `if (instr->opcode() == HloOpcode::kFusion) {`, after setting the stream id for kCopyStart, `AnnotateStreamAttributesForUsers` should take care of the kCopyDone?

Reply:
@Tixxx I tried that and found AnnotateStreamAttributesForUsers() inserted an async-done as the user of copy-start and crashed the  HloVerifier. "The operand of a copy-done instruction needs to be copy-start, found async-done". Can I avoid the generation of async-done?

Reply:
Are you talking about this pass StreamAttributeAsyncWrapper?  StreamAttributeAnnotator shouldnt add any additional async instructions, it only adds annotations. StreamAttributeAsyncWrapper will take the stream attribute and turn it into asyncstart/done. Hlo remat runs after StreamAttributeAsyncWrapper, so i dont see an issue there.